### PR TITLE
[WIP] Call behaviors with target methods.

### PR DESCRIPTION
**Pull request by Scooletz**
This replaces obtaining an interface methods with a target method of the instance.

### CODE REVIEW

**Comment by Scooletz** [comment]
@danielmarbach If you have a proper benchmark set, could you run this with it? It changes the called method from interface to the class, skipping interface dispatch as well as leaving some space for possible JIT optimizations (if any).

**Comment by danielmarbach** [comment]
I'm running it right now

**Comment by danielmarbach** [comment]
```ini

Host Process Environment Information:
BenchmarkDotNet=v0.9.8.0
OS=Microsoft Windows NT 6.2.9200.0
Processor=Intel(R) Core(TM) i7-4980HQ CPU 2.80GHz, ProcessorCount=8
Frequency=2728071 ticks, Resolution=366.5594 ns, Timer=TSC
CLR=MS.NET 4.0.30319.42000, Arch=64-bit RELEASE [RyuJIT]
GC=Concurrent Workstation
JitModules=clrjit-v4.6.1055.0

Type=PipelineExecution  Mode=SingleRun  LaunchCount=1  
WarmupCount=1  TargetCount=1  

```
                         Method |      GarbageCollection |   Calls | PipelineDepth |        Median |    StdDev | Scaled |    Gen 0 | Gen 1 | Gen 2 | Bytes Allocated/Op |
------------------------------- |----------------------- |-------- |-------------- |-------------- |---------- |------- |--------- |------ |------ |------------------- |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |   **20000** |            **10** |    **28.0029 ms** | **0.0000 ms** |   **1.00** |        **-** |     **-** |     **-** |      **56'066'397.33** |
  V6_PipelineAfterOptimizations |      Concurrent Server |   20000 |            10 |     2.6638 ms | 0.0000 ms |   0.10 |        - |     - |     - |         282'981.33 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |   20000 |            10 |    33.5200 ms | 0.0000 ms |   1.00 |     9.00 |     - |     - |      56'111'672.00 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |   20000 |            10 |     2.6444 ms | 0.0000 ms |   0.08 |        - |     - |     - |         320'042.67 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |   **20000** |            **20** |    **52.4059 ms** | **0.0000 ms** |   **1.00** |        **-** |     **-** |     **-** |     **109'983'437.33** |
  V6_PipelineAfterOptimizations |      Concurrent Server |   20000 |            20 |     5.4185 ms | 0.0000 ms |   0.10 |        - |     - |     - |         569'170.67 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |   20000 |            20 |    66.6962 ms | 0.0000 ms |   1.00 |    31.00 |     - |     - |     110'089'738.67 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |   20000 |            20 |     5.6197 ms | 0.0000 ms |   0.08 |        - |     - |     - |         606'432.00 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |   **20000** |            **40** |    **99.6844 ms** | **0.0000 ms** |   **1.00** |        **-** |     **-** |     **-** |     **217'986'461.33** |
  V6_PipelineAfterOptimizations |      Concurrent Server |   20000 |            40 |    12.4982 ms | 0.0000 ms |   0.13 |        - |     - |     - |       1'107'722.67 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |   20000 |            40 |   113.1543 ms | 0.0000 ms |   1.00 |    83.00 |     - |     - |     218'065'125.33 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |   20000 |            40 |    14.9124 ms | 0.0000 ms |   0.13 |        - |     - |     - |       1'142'461.33 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |   **40000** |            **10** |    **54.7086 ms** | **0.0000 ms** |   **1.00** |        **-** |     **-** |     **-** |     **111'740'082.67** |
  V6_PipelineAfterOptimizations |      Concurrent Server |   40000 |            10 |     5.7715 ms | 0.0000 ms |   0.11 |        - |     - |     - |         285'258.67 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |   40000 |            10 |    68.3567 ms | 0.0000 ms |   1.00 |    32.00 |     - |     - |     111'845'442.67 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |   40000 |            10 |     5.5427 ms | 0.0000 ms |   0.08 |        - |     - |     - |         320'042.67 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |   **40000** |            **20** |   **130.1630 ms** | **0.0000 ms** |   **1.00** |        **-** |     **-** |     **-** |     **219'414'008.00** |
  V6_PipelineAfterOptimizations |      Concurrent Server |   40000 |            20 |    10.8751 ms | 0.0000 ms |   0.08 |        - |     - |     - |         570'720.00 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |   40000 |            20 |   117.1927 ms | 0.0000 ms |   1.00 |    83.00 |     - |     - |     219'603'642.67 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |   40000 |            20 |    11.4088 ms | 0.0000 ms |   0.10 |        - |     - |     - |         606'432.00 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |   **40000** |            **40** |   **205.0196 ms** | **0.0000 ms** |   **1.00** |     **1.00** |     **-** |     **-** |     **434'974'717.33** |
  V6_PipelineAfterOptimizations |      Concurrent Server |   40000 |            40 |    22.2872 ms | 0.0000 ms |   0.11 |        - |     - |     - |       1'072'973.33 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |   40000 |            40 |   216.8147 ms | 0.0000 ms |   1.00 |   185.00 |     - |     - |     434'994'232.00 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |   40000 |            40 |    21.1769 ms | 0.0000 ms |   0.10 |        - |     - |     - |       1'142'461.33 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |   **80000** |            **10** |   **111.9905 ms** | **0.0000 ms** |   **1.00** |        **-** |     **-** |     **-** |     **223'259'280.00** |
  V6_PipelineAfterOptimizations |      Concurrent Server |   80000 |            10 |    11.3164 ms | 0.0000 ms |   0.10 |        - |     - |     - |         282'981.33 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |   80000 |            10 |   120.2014 ms | 0.0000 ms |   1.00 |    85.00 |     - |     - |     223'366'920.00 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |   80000 |            10 |    10.4638 ms | 0.0000 ms |   0.09 |        - |     - |     - |         320'042.67 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |   **80000** |            **20** |   **214.4607 ms** | **0.0000 ms** |   **1.00** |     **1.00** |     **-** |     **-** |     **438'481'576.00** |
  V6_PipelineAfterOptimizations |      Concurrent Server |   80000 |            20 |    21.9642 ms | 0.0000 ms |   0.10 |        - |     - |     - |         569'117.33 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |   80000 |            20 |   215.5552 ms | 0.0000 ms |   1.00 |   187.00 |     - |     - |     438'621'138.67 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |   80000 |            20 |    23.0111 ms | 0.0000 ms |   0.11 |        - |     - |     - |         606'432.00 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |   **80000** |            **40** |   **417.3689 ms** | **0.0000 ms** |   **1.00** |     **5.00** |     **-** |     **-** |     **868'891'186.67** |
  V6_PipelineAfterOptimizations |      Concurrent Server |   80000 |            40 |    45.6810 ms | 0.0000 ms |   0.11 |        - |     - |     - |       1'106'301.33 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |   80000 |            40 |   425.3170 ms | 0.0000 ms |   1.00 |   391.00 |     - |     - |     868'918'890.67 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |   80000 |            40 |    41.2409 ms | 0.0000 ms |   0.10 |        - |     - |     - |       1'142'461.33 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |  **160000** |            **10** |   **233.8227 ms** | **0.0000 ms** |   **1.00** |     **1.00** |     **-** |     **-** |     **446'314'725.33** |
  V6_PipelineAfterOptimizations |      Concurrent Server |  160000 |            10 |    21.8008 ms | 0.0000 ms |   0.09 |        - |     - |     - |         283'760.00 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |  160000 |            10 |   229.7187 ms | 0.0000 ms |   1.00 |   191.00 |     - |     - |     446'410'365.33 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |  160000 |            10 |    21.5563 ms | 0.0000 ms |   0.09 |        - |     - |     - |         320'042.67 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |  **160000** |            **20** |   **442.4302 ms** | **0.0000 ms** |   **1.00** |     **5.00** |     **-** |     **-** |     **876'478'837.33** |
  V6_PipelineAfterOptimizations |      Concurrent Server |  160000 |            20 |    43.9611 ms | 0.0000 ms |   0.10 |        - |     - |     - |         570'728.00 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |  160000 |            20 |   418.9986 ms | 0.0000 ms |   1.00 |   395.00 |     - |     - |     876'630'256.00 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |  160000 |            20 |    43.1547 ms | 0.0000 ms |   0.10 |        - |     - |     - |         606'432.00 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |  **160000** |            **40** |   **893.2718 ms** | **0.0000 ms** |   **1.00** |    **11.00** |     **-** |     **-** |   **1'736'614'997.33** |
  V6_PipelineAfterOptimizations |      Concurrent Server |  160000 |            40 |    85.2309 ms | 0.0000 ms |   0.10 |        - |     - |     - |       1'070'549.33 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |  160000 |            40 |   829.3010 ms | 0.0000 ms |   1.00 |   800.00 |  1.00 |     - |   1'736'733'264.00 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |  160000 |            40 |    87.4754 ms | 0.0000 ms |   0.11 |        - |     - |     - |       1'142'461.33 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |  **320000** |            **10** |   **460.3564 ms** | **0.0000 ms** |   **1.00** |     **5.00** |     **-** |     **-** |     **892'397'402.67** |
  V6_PipelineAfterOptimizations |      Concurrent Server |  320000 |            10 |    43.7016 ms | 0.0000 ms |   0.09 |        - |     - |     - |         284'285.33 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |  320000 |            10 |   454.4654 ms | 0.0000 ms |   1.00 |   403.00 |     - |     - |     892'497'013.33 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |  320000 |            10 |    42.6679 ms | 0.0000 ms |   0.09 |        - |     - |     - |         320'042.67 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |  **320000** |            **20** |   **874.7360 ms** | **0.0000 ms** |   **1.00** |    **10.00** |     **-** |     **-** |   **1'752'532'117.33** |
  V6_PipelineAfterOptimizations |      Concurrent Server |  320000 |            20 |    88.1902 ms | 0.0000 ms |   0.10 |        - |     - |     - |         569'170.67 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |  320000 |            20 |   913.1350 ms | 0.0000 ms |   1.00 |   809.00 |     - |     - |   1'752'680'434.67 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |  320000 |            20 |    88.8023 ms | 0.0000 ms |   0.10 |        - |     - |     - |         606'432.00 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |  **320000** |            **40** | **1,638.5637 ms** | **0.0000 ms** |   **1.00** |    **21.00** |     **-** |     **-** |   **3'472'311'280.00** |
  V6_PipelineAfterOptimizations |      Concurrent Server |  320000 |            40 |   167.5712 ms | 0.0000 ms |   0.10 |        - |     - |     - |       1'070'549.33 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |  320000 |            40 | 1,683.9470 ms | 0.0000 ms |   1.00 | 1,622.00 |  1.00 |     - |   3'472'322'869.33 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |  320000 |            40 |   178.2402 ms | 0.0000 ms |   0.11 |        - |     - |     - |       1'142'461.33 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |  **640000** |            **10** |   **896.0100 ms** | **0.0000 ms** |   **1.00** |    **11.00** |     **-** |     **-** |   **1'784'545'922.67** |
  V6_PipelineAfterOptimizations |      Concurrent Server |  640000 |            10 |    91.0031 ms | 0.0000 ms |   0.10 |        - |     - |     - |         318'570.67 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |  640000 |            10 |   928.9524 ms | 0.0000 ms |   1.00 |   825.00 |     - |     - |   1'784'666'285.33 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |  640000 |            10 |    93.4162 ms | 0.0000 ms |   0.10 |        - |     - |     - |         320'042.67 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |  **640000** |            **20** | **1,724.4720 ms** | **0.0000 ms** |   **1.00** |    **22.00** |     **-** |     **-** |   **3'504'546'210.67** |
  V6_PipelineAfterOptimizations |      Concurrent Server |  640000 |            20 |   173.3533 ms | 0.0000 ms |   0.10 |        - |     - |     - |         569'117.33 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |  640000 |            20 | 1,642.4001 ms | 0.0000 ms |   1.00 | 1,639.00 |  1.00 |     - |   3'504'808'602.67 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |  640000 |            20 |   176.4613 ms | 0.0000 ms |   0.11 |        - |     - |     - |         606'432.00 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** |  **640000** |            **40** | **3,437.6550 ms** | **0.0000 ms** |   **1.00** |    **43.00** |     **-** |     **-** |   **6'943'461'029.33** |
  V6_PipelineAfterOptimizations |      Concurrent Server |  640000 |            40 |   337.1206 ms | 0.0000 ms |   0.10 |        - |     - |     - |       1'071'640.00 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation |  640000 |            40 | 3,384.3555 ms | 0.0000 ms |   1.00 | 3,265.00 |  2.00 |     - |   6'943'546'992.00 |
  V6_PipelineAfterOptimizations | Concurrent Workstation |  640000 |            40 |   392.0833 ms | 0.0000 ms |   0.12 |        - |     - |     - |       1'142'461.33 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** | **1280000** |            **10** | **1,820.3844 ms** | **0.0000 ms** |   **1.00** |    **22.00** |     **-** |     **-** |   **3'568'832'429.33** |
  V6_PipelineAfterOptimizations |      Concurrent Server | 1280000 |            10 |   180.6335 ms | 0.0000 ms |   0.10 |        - |     - |     - |         282'949.33 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation | 1280000 |            10 | 1,725.1388 ms | 0.0000 ms |   1.00 | 1,672.00 |     - |     - |   3'569'013'229.33 |
  V6_PipelineAfterOptimizations | Concurrent Workstation | 1280000 |            10 |   228.6484 ms | 0.0000 ms |   0.13 |        - |     - |     - |         320'042.67 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** | **1280000** |            **20** | **3,513.0457 ms** | **0.0000 ms** |   **1.00** |    **44.00** |     **-** |     **-** |   **7'008'756'120.00** |
  V6_PipelineAfterOptimizations |      Concurrent Server | 1280000 |            20 |   360.9441 ms | 0.0000 ms |   0.10 |        - |     - |     - |         570'450.67 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation | 1280000 |            20 | 3,324.1334 ms | 0.0000 ms |   1.00 | 3,298.00 |  2.00 |     - |   7'009'042'114.67 |
  V6_PipelineAfterOptimizations | Concurrent Workstation | 1280000 |            20 |   354.8665 ms | 0.0000 ms |   0.11 |        - |     - |     - |         606'432.00 |
 **V6_PipelineBeforeOptimizations** |      **Concurrent Server** | **1280000** |            **40** | **6,560.9539 ms** | **0.0000 ms** |   **1.00** |    **91.00** |     **-** |     **-** |  **13'885'940'568.00** |
  V6_PipelineAfterOptimizations |      Concurrent Server | 1280000 |            40 |   722.0424 ms | 0.0000 ms |   0.11 |        - |     - |     - |       1'106'853.33 |
 V6_PipelineBeforeOptimizations | Concurrent Workstation | 1280000 |            40 | 6,561.0862 ms | 0.0000 ms |   1.00 | 6,548.00 |  7.00 |     - |  13'886'051'664.00 |
  V6_PipelineAfterOptimizations | Concurrent Workstation | 1280000 |            40 |   694.0908 ms | 0.0000 ms |   0.11 |        - |     - |     - |       1'142'461.33 |

File: src/NServiceBus.Core/Pipeline/PipelineExecutionExtensions.cs [hunk 1]
```diff
@@ -38,7 +38,7 @@ public static Delegate CreatePipelineExecutionExpression(this IBehavior[] behavi
                 {
                     throw new InvalidOperationException("Behaviors must implement IBehavior<TInContext, TOutContext>");
                 }
-                var methodInfo = behaviorInterfaceType.GetMethods().FirstOrDefault();
+                var methodInfo = currentBehavior.GetType().GetInterfaceMap(behaviorInterfaceType).TargetMethods.FirstOrDefault();
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Pipeline/PipelineExecutionExtensions.cs]: Cool stuff

**Comment by Scooletz** [comment]
I compared this with the results of your PR and I can't see a real difference. Feel free to drop this (or accept) :)

**Comment by danielmarbach** [comment]
I think it is a good change. I'll take it in next week

**Comment by danielmarbach** [comment]
Pulled into https://github.com/Particular/NServiceBus/pull/4125/commits/02b8e607279b9cac76965be51a51319209b4af02

**Pull request closed without merging**